R/model.R

```diff
@@ -914,6 +914,180 @@ sample_method <- function(data = NULL,
+#'   Linux and clang on macOS.
+#' 
+#'   In the call to the `$mpi_sample()` method, we can additionally provide
+#'   the name of the MPI launcher (`mpi_cmd`), which defaults to "mpiexec",
+#'   and any other MPI launch arguments. In most cases, it is enough to
+#'   only define the number of processes with `mpi_args = list("n" = 4)`.
+#' 
+#'   An example of a call of `$mpi_sample()`:
+#'   ```
+#'   cpp_options = list(STAN_MPI = TRUE, CXX="mpicxx", TBB_CXX_TYPE="gcc")
+#'   fit <- mod$mpi_sample(data_list, mpi_args = c("-n", 4))
+#'   ```
+#'
+#' @section Usage:
+#'   ```
+#'   $mpi_sample(
+#'     data = NULL,
+#'     mpi_cmd = "mpiexec",
+#'     mpi_args = NULL,
+#'     seed = NULL,
+#'     refresh = NULL,
+#'     init = NULL,
+#'     save_latent_dynamics = FALSE,
+#'     output_dir = NULL,
+#'     chains = 4,
+#'     parallel_chains = getOption("mc.cores", 1),
+#'     chain_ids = seq_len(chains),
+#'     iter_warmup = NULL,
+#'     iter_sampling = NULL,
+#'     save_warmup = FALSE,
+#'     thin = NULL,
+#'     max_treedepth = NULL,
+#'     adapt_engaged = TRUE,
+#'     adapt_delta = NULL,
+#'     step_size = NULL,
+#'     metric = NULL,
+#'     metric_file = NULL,
+#'     inv_metric = NULL,
+#'     init_buffer = NULL,
+#'     term_buffer = NULL,
+#'     window = NULL,
+#'     fixed_param = FALSE,
+#'     sig_figs = NULL,
+#'     validate_csv = TRUE,
+#'     show_messages = TRUE
+#'   )
+#'   ```
+#'
+#' @section Arguments:
+#'   * `mpi_cmd`: (character vector) The MPI launcher used for launching MPI processes.
+#'     The default launcher is `mpiexec`.
+#'   * `mpi_args`: (list) A list of arguments to use when launching MPI processes.
+#'     For example, mpi_args = list("n" = 4) launches the executable as
+#'     `mpiexec -n 4 model_executable`, followed by CmdStan arguments
+#'     for the model executable.
+#'   * `data`, `seed`, `refresh`, `init`, `save_latent_dynamics`, `output_dir`,
+#'     `chains`, `parallel_chains`, `chain_ids`, `iter_warmup`, `iter_sampling`,
+#'     `save_warmup`, `thin`, `max_treedepth`, `adapt_engaged`, `adapt_delta`,
+#'     `step_size`, `metric`, `metric_file`, `inv_metric`, `init_buffer`,
+#'     `term_buffer`, `window`, `fixed_param`, `sig_figs`, `validate_csv`,
+#'     `show_messages`:
+#'      Same as for the [`$sample()`][model-method-sample] method.
+#'
+#' @section Value: The `$mpi_sample()` method returns a [`CmdStanMCMC`] object.
+#'
+#' @template seealso-docs
+#' @inherit cmdstan_model examples
+#'
+NULL
+mpi_sample_method <- function(data = NULL,
+                          mpi_cmd = "mpiexec",
+                          mpi_args = NULL,
+                          seed = NULL,
+                          refresh = NULL,
+                          init = NULL,
+                          save_latent_dynamics = FALSE,
+                          output_dir = NULL,
+                          chains = 1,
+                          parallel_chains = getOption("mc.cores", 1),
```

Comment:
The default for chains in `$mpi_sample()` is 1, while for `$sample()` its 4. Or should we leave it the same as for `$sample()`?

Reply:
what happens when `chains=4` in terms of distributing processes? With `mpiexec -n 4` is each chain solved by 1 process?

Reply:
If it's `chains = 4, parallel_chains = 1` that is the same as 4 sequential $mpi_sample(chains = 1) and this is just a convenience so the draws are merged together in the fit.

If `chains = 4, parallel_chains = 4` that means 4 mpiexec calls with n=4. Not sure that is useful or if we should just fix parallel_chains to 1.

Reply:
I'm confused, as I thought `mpi_args=c("-n", 4...)` controls the total number of MPI processes. But looks like `parallel_chains=4` implies `mpiexec -n 4` too. Is that right?

Reply:
Without MPI, parallel_chains just means that 4 "./model args" calls are made and 4 models processes run in parallel.
So like running this in shell
```
for i in {1..4}
    do
      ./bernoulli sample data file=bernoulli.data.json \
      output file=output_${i}.csv &
    done
```

In `$mpi_sample()` this would mean 4 processes that would run "mpiexec -n x ./model args". Does that make more sense?

Reply:
As I said, I am not sure this is something we need.

Reply:
> 4 processes that would run mpiexec -n x ./model args

Is this equivalent to
```bash
for i in {1..4}
    do
      mpiexec -n x ./bernoulli sample data file=bernoulli.data.json \
      output file=output_${i}.csv &
    done
```
?

Reply:
Yes.

Reply:
Ok, thx. What does `chains = 1, parallel_chains = 4` amount to?

Reply:
Falls back to `chains = 1, parallel_chains = 1` silently. 
`chains = 4, parallel_chains = 2` runs at most 2 chains at one time. Once 1 finishes it can be replaced by another.

Reply:
`parallel_chains` is kind of like cores in `rstan`, we just thought that `cores` is not the best name for it. More so if you add in within-chain threading.

Reply:
I find this really confusing, as arg `chains` and `parallel_chains` seem to have overlapping meanings. What happens if we remove `paralle_chains` and only use `chains` in `mpi_sample`?

Reply:
This comes mainly from non-parallel MCMC sampling. For example users want to run 4 chains but want to only use 2 cores for two chains and use the other two for something else. rstan uses the `cores` argument for this same thing. 

I think this makes less sense or is less useful in the context of within-chain parallelization with threading or MPI, because if someone uses parallelization its likely they want all the CPU power.

We can remove parallel_chains, we just need to decide what to do in the case of chains > 1. Do we run chains sequentially (paralllel_chains = 1) or run all of them at once (parallel_chains = chains).
My gut feeling is that for we go with the former?

Reply:
My assumption is just that one typically runs a single `mpiexec` with maximum n?

Reply:
It depends on the context. For now I agree with you semantically it makes more sense to have `chains=4 + mpi_arg("-n", x)` equivalent to
```bash
for i in {1..4}
    do
      mpiexec -n x ./bernoulli sample data file=bernoulli.data.json \
      output file=output_${i}.csv &
    done
```

Reply:
Ok, let's sleep on this a bit more.